cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/DefaultArtifactRepository.java

```diff
@@ -97,16 +98,24 @@ public DefaultArtifactRepository(CConfiguration cConf, ArtifactStore artifactSto
     this.artifactStore = artifactStore;
     this.artifactClassLoaderFactory = new ArtifactClassLoaderFactory(cConf, programRunnerFactory);
     this.artifactInspector = new ArtifactInspector(cConf, artifactClassLoaderFactory);
-    this.systemArtifactDirs = new ArrayList<>();
+    this.systemArtifactDirs = new HashSet<>();
     String systemArtifactsDir = cConf.get(Constants.AppFabric.SYSTEM_ARTIFACTS_DIR);
     if (!Strings.isNullOrEmpty(systemArtifactsDir)) {
+      String sparkDirStr = SparkCompat.get(cConf).getCompat();
+
       for (String dir : systemArtifactsDir.split(";")) {
         File file = new File(dir);
         if (!file.isDirectory()) {
           LOG.warn("Ignoring {} because it is not a directory.", file);
           continue;
         }
         systemArtifactDirs.add(file);
+        // if users have a non-default value for system artifacts, they may be missing the spark compat directory
```

Comment:
Even the non-default value for system artifacts needs this now (not just non-default values).
So this comment is somewhat inaccurate / misleading.